# Difference method

In this model, we will use what we called the Difference method. We first calculate the difference between the experimental band gap and the PBE band gap, and we then evaluate the error on this difference.

In [1]:
def setup_threading():
    import os
    os.environ['OPENBLAS_NUM_THREADS'] = '1'
    os.environ['MKL_NUM_THREADS'] = '1'
    os.environ["OMP_NUM_THREADS"] = "1"
    os.environ["TF_NUM_INTRAOP_THREADS"] = "1"
    os.environ["TF_NUM_INTEROP_THREADS"] = "1"
    os.environ["CUDA_VISIBLE_DEVICES"] = "-1"

In [2]:
setup_threading()

In [3]:
from modnet.preprocessing import MODData
from modnet.models.vanilla import MODNetModel
from modnet.hyper_opt.fit_genetic import FitGenetic
import numpy as np
import os
import copy
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

In [4]:
from sklearn.model_selection import KFold
from modnet.preprocessing import MODData

def shuffle_MD(data,random_state=10):
    data = copy.deepcopy(data)
    ids = data.df_targets.sample(frac=1,random_state=random_state).index
    data.df_featurized = data.df_featurized.reindex([ids])
    data.df_targets = data.df_targets.reindex([ids])
    data.df_structure = data.df_structure.reindex([ids])
    
    return data

def MDKsplit(data,n_splits=5,random_state=10):
    #data = shuffle_MD(data,random_state=random_state)
    ids = np.array(data.structure_ids)
    kf = KFold(n_splits=n_splits,shuffle=True,random_state=random_state)
    folds = []
    for train_idx, val_idx in kf.split(ids):
        data_train = MODData(data.df_structure.iloc[train_idx]['structure'].values,data.df_targets.iloc[train_idx].values,target_names=data.df_targets.columns,structure_ids=ids[train_idx])
        data_train.df_featurized = data.df_featurized.iloc[train_idx]
        #data_train.optimal_features = data.optimal_features
        
        data_val = MODData(data.df_structure.iloc[val_idx]['structure'].values,data.df_targets.iloc[val_idx].values,target_names=data.df_targets.columns,structure_ids=ids[val_idx])
        data_val.df_featurized = data.df_featurized.iloc[val_idx]
        #data_val.optimal_features = data.optimal_features

        folds.append((data_train,data_val))
        
    return folds

def MD_append(md,lmd):
    md = copy.deepcopy(md)
    for m in lmd:
        md.df_structure.append(m.df_structure)
        md.df_targets.append(m.df_targets)
        md.df_featurized.append(m.df_featurized)
    return md

In [5]:
md_exp = MODData.load('exp_gap_all_mpid')
md_exp.df_targets.columns = ['exp_gap','mp_id']
md_pbe = MODData.load('pbe_gap.zip')
md_pbe.df_targets.columns = ['gap']
md_joint = MODData.load('exp_pbe_joint')


If you use the ChemEnv tool for your research, please consider citing the following reference(s) :
David Waroquiers, Xavier Gonze, Gian-Marco Rignanese, Cathrin Welker-Nieuwoudt, Frank Rosowski,
Michael Goebel, Stephan Schenk, Peter Degelmann, Rute Andre, Robert Glaum, and Geoffroy Hautier,
"Statistical analysis of coordination environments in oxides",
Chem. Mater., 2017, 29 (19), pp 8346-8360,
DOI: 10.1021/acs.chemmater.7b02766

2021-06-14 17:44:29,521 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f3f28068970> object, created with modnet version <=0.1.7
2021-06-14 17:44:37,814 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f3eb95a29d0> object, created with modnet version <=0.1.7
2021-06-14 17:44:38,912 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f3f280680d0> object, created with modnet version <=0.1.7


In [6]:
md_joint.df_targets

exp_gap  pbe_gap  difference
mp-12699      0.00   0.0000      0.0000
mp-559459     3.40   2.1704      1.2296
mp-21162      0.00   0.0000      0.0000
mp-1306       0.00   0.0000      0.0000
mp-15252      0.53   0.1049      0.4251
...            ...      ...         ...
mp-556541     1.80   1.5135      0.2865
mp-2602       1.04   0.0000      1.0400
mp-30366      0.00   0.0000      0.0000
mp-3718       1.45   0.2418      1.2082
mp-3709       0.00   0.0000      0.0000

[1703 rows x 3 columns]

In [7]:
md_joint.df_targets['difference'] = md_joint.df_targets['exp_gap'] - md_joint.df_targets['pbe_gap']

In [8]:
md_joint.df_targets

exp_gap  pbe_gap  difference
mp-12699      0.00   0.0000      0.0000
mp-559459     3.40   2.1704      1.2296
mp-21162      0.00   0.0000      0.0000
mp-1306       0.00   0.0000      0.0000
mp-15252      0.53   0.1049      0.4251
...            ...      ...         ...
mp-556541     1.80   1.5135      0.2865
mp-2602       1.04   0.0000      1.0400
mp-30366      0.00   0.0000      0.0000
mp-3718       1.45   0.2418      1.2082
mp-3709       0.00   0.0000      0.0000

[1703 rows x 3 columns]

In [9]:
md_joint.df_structure

structure
id                                                          
mp-12699   [[ 0.13616363 -2.60631371  1.50340165] Ca, [-1...
mp-559459  [[2.14049282 6.43464803 1.76199846] Cs, [-0.70...
mp-21162   [[0. 0. 0.] Sr, [0.        2.5259875 2.5259875...
mp-1306    [[3.259129   1.88166243 1.15971025] Th, [ 3.25...
mp-15252   [[0.95531657 1.042239   0.86229681] Cu, [2.803...
...                                                      ...
mp-556541  [[0.         6.49363948 5.82657217] In, [ 0.  ...
mp-2602     [[3.023318 3.023318 3.023318] Nd, [0. 0. 0.] As]
mp-30366   [[4.51774558 1.37529211 4.32659015] Ca, [1.109...
mp-3718    [[2.0697935  0.         1.28932533] Th, [0.   ...
mp-3709    [[0. 0. 0.] Tb, [0.       1.983905 2.537711] C...

[1703 rows x 1 columns]

In [10]:
from sklearn.metrics import mean_absolute_error
mean_absolute_error(md_joint.df_targets['exp_gap'], md_joint.df_targets['pbe_gap'])

0.429783975697596

In [11]:
md_joint.df_targets['difference'].mad()

0.5494667780829975

In [12]:
md_joint.df_targets = md_joint.df_targets.drop(columns=['exp_gap', 'pbe_gap'])

In [14]:
import uncertainty_toolbox as uct

k = 5
random_state = 202010
folds = MDKsplit(md_joint,n_splits=k,random_state=random_state)
errors = np.ones((int(len(md_joint.df_targets)/k),5))
maes = np.ones(5)
uncertainties = np.ones(5)
metricss = []
for i,f in enumerate(folds):
    train = f[0]
    test = f[1]
    #train.feature_selection(n=-1, use_precomputed_cross_nmi=True)
    fpath = 'train_joint_exp-pbe_{}_{}'.format(random_state,i+1)
    if os.path.exists(fpath):
        train = MODData.load(fpath)
    else:
        train.feature_selection(n=-1, use_precomputed_cross_nmi=True)
        train.save(fpath)
       
    # assure no overlap
    assert len(set(train.df_targets.index).intersection(set(test.df_targets.index))) == 0
    
    # difference prediction
    ga = FitGenetic(train)
    model = ga.run(size_pop=20, num_generations=10, n_jobs=10)
    
    pred, std = model.predict(test, return_unc=True)
    true = test.df_targets
    metrics = uct.metrics.get_all_metrics(pred['difference'].values, std['difference'].values, true['difference'].values)
    error = pred-true
    error = error['difference'].drop(pred.index[((pred['difference']).abs()>20)]) # drop unrealistic values: happens extremely rarely
    std = std['difference'].drop(std.index[((std['difference']).abs()>20)]) # drop unrealistic values: happens extremely rarely
    mae = np.abs(error.values).mean()
    uncertainty = np.abs(std).mean()
    print('MAE =')
    print(mae)
    print('uncertainty =')
    print(uncertainty)
    maes[i] = mae
    uncertainties[i] = uncertainty
    metricss.append(metrics)

2021-06-15 00:06:47,030 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,059 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,073 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,095 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,112 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,151 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,168 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,219 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,235 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,281 - modnet - INFO - Loaded DeBreuck2020Featurizer featurizer.
2021-06-15 00:06:47,858 - modnet - INFO - Loaded <modnet.preprocessing.MODData object at 0x7f3ea6d8e970> object, created with modnet version 0.1.9
2021-06-15 00

100%|██████████| 100/100 [17:12<00:00, 10.32s/it]

2021-06-15 00:24:03,746 - modnet - INFO - Loss per individual: ind 0: 0.939 	ind 1: 0.931 	ind 2: 0.888 	ind 3: 0.818 	ind 4: 0.872 	ind 5: 1.012 	ind 6: 0.927 	ind 7: 0.939 	ind 8: 1.204 	ind 9: 0.957 	ind 10: 0.921 	ind 11: 1.052 	ind 12: 0.936 	ind 13: 0.957 	ind 14: 0.988 	ind 15: 1.060 	ind 16: 1.088 	ind 17: 0.894 	ind 18: 0.881 	ind 19: 0.999 	


2021-06-15 00:24:05,297 - modnet - INFO - Generation number 1
2021-06-15 00:24:08,483 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [16:32<00:00,  9.93s/it]

2021-06-15 00:40:41,237 - modnet - INFO - Loss per individual: ind 0: 0.874 	ind 1: 1.052 	ind 2: 0.937 	ind 3: 0.864 	ind 4: 1.098 	ind 5: 0.880 	ind 6: 1.240 	ind 7: 0.983 	ind 8: 0.989 	ind 9: 0.960 	ind 10: 0.943 	ind 11: 0.858 	ind 12: 1.021 	ind 13: 0.904 	ind 14: 0.874 	ind 15: 0.929 	ind 16: 0.894 	ind 17: 1.051 	ind 18: 0.894 	ind 19: 0.865 	


2021-06-15 00:40:43,086 - modnet - INFO - Generation number 2
2021-06-15 00:40:46,976 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [17:09<00:00, 10.29s/it]

2021-06-15 00:57:56,341 - modnet - INFO - Loss per individual: ind 0: 0.941 	ind 1: 0.955 	ind 2: 0.910 	ind 3: 0.995 	ind 4: 0.957 	ind 5: 1.206 	ind 6: 0.840 	ind 7: 0.863 	ind 8: 0.877 	ind 9: 0.893 	ind 10: 1.008 	ind 11: 0.938 	ind 12: 0.841 	ind 13: 0.957 	ind 14: 0.924 	ind 15: 0.928 	ind 16: 0.886 	ind 17: 0.979 	ind 18: 0.924 	ind 19: 0.882 	


2021-06-15 00:57:58,146 - modnet - INFO - Generation number 3
2021-06-15 00:58:02,153 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [10:01<00:00,  6.01s/it]

2021-06-15 01:08:03,354 - modnet - INFO - Loss per individual: ind 0: 0.978 	ind 1: 0.886 	ind 2: 0.888 	ind 3: 0.919 	ind 4: 0.918 	ind 5: 0.921 	ind 6: 0.858 	ind 7: 1.100 	ind 8: 0.870 	ind 9: 0.968 	ind 10: 0.982 	ind 11: 0.871 	ind 12: 0.865 	ind 13: 0.872 	ind 14: 0.960 	ind 15: 0.887 	ind 16: 1.081 	ind 17: 0.932 	ind 18: 0.882 	ind 19: 1.132 	


2021-06-15 01:08:04,454 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


 20%|██        | 2/10 [00:00<00:00, 10.26it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


 20%|██        | 2/10 [00:00<00:00, 10.47it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:01<00:00,  9.92it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.339
  RMSE          0.686
  MDAE          0.063
  MARPD         148.135
  R2            0.384
  Correlation   0.638
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.101
  Mean-absolute Calibration Error       0.084
  Miscalibration Area                   0.085
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.135
     Group Size: 0.56 -- Calibration Error: 0.098
     Group Size: 1.00 -- Calibration Error: 0.084
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.165
     Group Size: 0.56 -- Calibration Error: 0.117
     Group Size: 1.00 -- Calibration Error: 0.101
==============

100%|██████████| 100/100 [07:55<00:00,  4.76s/it]

2021-06-15 01:16:11,187 - modnet - INFO - Loss per individual: ind 0: 0.964 	ind 1: 104.485 	ind 2: 1.088 	ind 3: 1.044 	ind 4: 7.740 	ind 5: 0.924 	ind 6: 2.111 	ind 7: 25.589 	ind 8: 3.396 	ind 9: 89.341 	ind 10: 12.746 	ind 11: 5.877 	ind 12: 5.079 	ind 13: 36.875 	ind 14: 10.228 	ind 15: 0.969 	ind 16: 3.560 	ind 17: 2.058 	ind 18: 2.136 	ind 19: 10.810 	


2021-06-15 01:16:12,213 - modnet - INFO - Generation number 1
2021-06-15 01:16:14,557 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:22<00:00,  3.23s/it]

2021-06-15 01:21:37,480 - modnet - INFO - Loss per individual: ind 0: 0.925 	ind 1: 0.924 	ind 2: 1.155 	ind 3: 0.991 	ind 4: 38.751 	ind 5: 0.957 	ind 6: 0.977 	ind 7: 11.178 	ind 8: 1.106 	ind 9: 0.948 	ind 10: 0.955 	ind 11: 1.047 	ind 12: 1.028 	ind 13: 0.958 	ind 14: 1.099 	ind 15: 5.357 	ind 16: 0.993 	ind 17: 0.961 	ind 18: 1.049 	ind 19: 0.941 	


2021-06-15 01:21:38,349 - modnet - INFO - Generation number 2
2021-06-15 01:21:40,675 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:04<00:00,  3.04s/it]

2021-06-15 01:26:44,873 - modnet - INFO - Loss per individual: ind 0: 1.136 	ind 1: 1.164 	ind 2: 0.932 	ind 3: 0.944 	ind 4: 0.968 	ind 5: 1.001 	ind 6: 0.933 	ind 7: 1.005 	ind 8: 0.989 	ind 9: 1.048 	ind 10: 0.977 	ind 11: 0.936 	ind 12: 0.966 	ind 13: 1.014 	ind 14: 0.947 	ind 15: 1.040 	ind 16: 0.925 	ind 17: 0.975 	ind 18: 0.967 	ind 19: 1.030 	


2021-06-15 01:26:45,684 - modnet - INFO - Generation number 3
2021-06-15 01:26:48,154 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:07<00:00,  3.07s/it]

2021-06-15 01:31:55,622 - modnet - INFO - Loss per individual: ind 0: 0.964 	ind 1: 1.025 	ind 2: 1.008 	ind 3: 1.027 	ind 4: 0.979 	ind 5: 1.011 	ind 6: 0.952 	ind 7: 0.962 	ind 8: 0.940 	ind 9: 0.994 	ind 10: 0.976 	ind 11: 0.988 	ind 12: 0.940 	ind 13: 0.965 	ind 14: 0.987 	ind 15: 0.973 	ind 16: 6.344 	ind 17: 0.982 	ind 18: 0.937 	ind 19: 1.025 	


2021-06-15 01:31:56,433 - modnet - INFO - Generation number 4
2021-06-15 01:32:07,527 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:54<00:00,  3.55s/it]

2021-06-15 01:38:02,528 - modnet - INFO - Loss per individual: ind 0: 0.996 	ind 1: 1.053 	ind 2: 0.966 	ind 3: 1.054 	ind 4: 0.939 	ind 5: 1.052 	ind 6: 1.002 	ind 7: 1.037 	ind 8: 0.932 	ind 9: 0.922 	ind 10: 1.023 	ind 11: 0.994 	ind 12: 0.986 	ind 13: 0.919 	ind 14: 14.004 	ind 15: 0.965 	ind 16: 0.964 	ind 17: 1.017 	ind 18: 0.932 	ind 19: 0.934 	


2021-06-15 01:38:03,357 - modnet - INFO - Generation number 5
2021-06-15 01:38:06,355 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:38<00:00,  3.38s/it]

2021-06-15 01:43:44,801 - modnet - INFO - Loss per individual: ind 0: 0.990 	ind 1: 0.973 	ind 2: 1.045 	ind 3: 0.920 	ind 4: 0.962 	ind 5: 0.928 	ind 6: 0.947 	ind 7: 1.021 	ind 8: 1.001 	ind 9: 0.943 	ind 10: 0.992 	ind 11: 20.777 	ind 12: 1.015 	ind 13: 0.940 	ind 14: 1.007 	ind 15: 3.987 	ind 16: 0.965 	ind 17: 0.978 	ind 18: 1.012 	ind 19: 0.972 	


2021-06-15 01:43:45,621 - modnet - INFO - Generation number 6
2021-06-15 01:43:48,919 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:01<00:00,  3.61s/it]

2021-06-15 01:49:50,063 - modnet - INFO - Loss per individual: ind 0: 1.016 	ind 1: 0.932 	ind 2: 0.938 	ind 3: 0.931 	ind 4: 0.959 	ind 5: 1.006 	ind 6: 0.972 	ind 7: 3.933 	ind 8: 0.951 	ind 9: 0.926 	ind 10: 1.000 	ind 11: 1.039 	ind 12: 0.950 	ind 13: 1.029 	ind 14: 0.940 	ind 15: 0.984 	ind 16: 0.974 	ind 17: 0.933 	ind 18: 0.954 	ind 19: 0.965 	


2021-06-15 01:49:50,902 - modnet - INFO - Generation number 7
2021-06-15 01:49:54,517 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:48<00:00,  3.49s/it]

2021-06-15 01:55:43,455 - modnet - INFO - Loss per individual: ind 0: 0.934 	ind 1: 1.022 	ind 2: 0.943 	ind 3: 0.947 	ind 4: 1.081 	ind 5: 0.960 	ind 6: 1.033 	ind 7: 1.044 	ind 8: 0.947 	ind 9: 0.950 	ind 10: 0.943 	ind 11: 0.991 	ind 12: 0.935 	ind 13: 0.968 	ind 14: 0.994 	ind 15: 0.936 	ind 16: 0.978 	ind 17: 0.967 	ind 18: 1.066 	ind 19: 0.993 	


2021-06-15 01:55:44,273 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


 20%|██        | 2/10 [00:00<00:00, 11.35it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


 20%|██        | 2/10 [00:00<00:00, 11.14it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:00<00:00, 10.85it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.292
  RMSE          0.648
  MDAE          0.050
  MARPD         146.950
  R2            0.358
  Correlation   0.620
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.043
  Mean-absolute Calibration Error       0.033
  Miscalibration Area                   0.033
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.103
     Group Size: 0.56 -- Calibration Error: 0.052
     Group Size: 1.00 -- Calibration Error: 0.033
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.118
     Group Size: 0.56 -- Calibration Error: 0.056
     Group Size: 1.00 -- Calibration Error: 0.043
==============

100%|██████████| 100/100 [05:07<00:00,  3.08s/it]

2021-06-15 02:01:02,756 - modnet - INFO - Loss per individual: ind 0: 0.934 	ind 1: 1.151 	ind 2: 2.004 	ind 3: 0.915 	ind 4: 1.034 	ind 5: 6.484 	ind 6: 0.876 	ind 7: 1.230 	ind 8: 0.892 	ind 9: 0.932 	ind 10: 1.831 	ind 11: 1.075 	ind 12: 0.930 	ind 13: 0.931 	ind 14: 1.148 	ind 15: 11.415 	ind 16: 1.531 	ind 17: 4.205 	ind 18: 13.940 	ind 19: 0.944 	


2021-06-15 02:01:03,542 - modnet - INFO - Generation number 1
2021-06-15 02:01:06,171 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:53<00:00,  2.93s/it]

2021-06-15 02:05:59,595 - modnet - INFO - Loss per individual: ind 0: 1.199 	ind 1: 0.902 	ind 2: 0.859 	ind 3: 0.961 	ind 4: 0.876 	ind 5: 0.910 	ind 6: 4.737 	ind 7: 0.892 	ind 8: 0.931 	ind 9: 0.909 	ind 10: 0.912 	ind 11: 0.889 	ind 12: 0.927 	ind 13: 0.879 	ind 14: 0.857 	ind 15: 0.880 	ind 16: 0.908 	ind 17: 1.050 	ind 18: 1.003 	ind 19: 0.961 	


2021-06-15 02:06:00,388 - modnet - INFO - Generation number 2
2021-06-15 02:06:03,231 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:40<00:00,  3.41s/it]

2021-06-15 02:11:44,109 - modnet - INFO - Loss per individual: ind 0: 0.861 	ind 1: 0.933 	ind 2: 0.945 	ind 3: 1.071 	ind 4: 0.875 	ind 5: 0.891 	ind 6: 0.948 	ind 7: 1.045 	ind 8: 0.908 	ind 9: 1.028 	ind 10: 0.921 	ind 11: 0.902 	ind 12: 0.896 	ind 13: 0.949 	ind 14: 0.911 	ind 15: 1.009 	ind 16: 0.903 	ind 17: 0.924 	ind 18: 0.935 	ind 19: 10.139 	


2021-06-15 02:11:44,915 - modnet - INFO - Generation number 3
2021-06-15 02:11:47,820 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:45<00:00,  2.85s/it]

2021-06-15 02:16:33,190 - modnet - INFO - Loss per individual: ind 0: 0.919 	ind 1: 0.924 	ind 2: 0.933 	ind 3: 0.896 	ind 4: 0.863 	ind 5: 0.920 	ind 6: 0.882 	ind 7: 0.924 	ind 8: 0.905 	ind 9: 0.916 	ind 10: 0.915 	ind 11: 0.879 	ind 12: 0.933 	ind 13: 0.885 	ind 14: 0.956 	ind 15: 0.881 	ind 16: 0.870 	ind 17: 0.904 	ind 18: 0.894 	ind 19: 0.949 	


2021-06-15 02:16:34,029 - modnet - INFO - Generation number 4
2021-06-15 02:16:37,099 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:31<00:00,  2.72s/it]

2021-06-15 02:21:08,671 - modnet - INFO - Loss per individual: ind 0: 0.911 	ind 1: 0.909 	ind 2: 0.901 	ind 3: 0.852 	ind 4: 0.912 	ind 5: 0.879 	ind 6: 0.952 	ind 7: 0.885 	ind 8: 0.875 	ind 9: 0.931 	ind 10: 0.886 	ind 11: 1.074 	ind 12: 0.864 	ind 13: 0.901 	ind 14: 0.882 	ind 15: 1.299 	ind 16: 0.905 	ind 17: 3.390 	ind 18: 0.900 	ind 19: 0.926 	


2021-06-15 02:21:09,488 - modnet - INFO - Generation number 5
2021-06-15 02:21:12,757 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:17<00:00,  2.58s/it]

2021-06-15 02:25:30,765 - modnet - INFO - Loss per individual: ind 0: 0.877 	ind 1: 0.892 	ind 2: 0.923 	ind 3: 1.022 	ind 4: 0.952 	ind 5: 0.934 	ind 6: 0.934 	ind 7: 0.900 	ind 8: 0.869 	ind 9: 0.929 	ind 10: 0.994 	ind 11: 0.915 	ind 12: 0.878 	ind 13: 0.957 	ind 14: 0.902 	ind 15: 0.888 	ind 16: 0.883 	ind 17: 0.902 	ind 18: 0.899 	ind 19: 0.920 	


2021-06-15 02:25:31,562 - modnet - INFO - Generation number 6
2021-06-15 02:25:35,132 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:53<00:00,  2.94s/it]

2021-06-15 02:30:29,055 - modnet - INFO - Loss per individual: ind 0: 0.884 	ind 1: 0.890 	ind 2: 0.912 	ind 3: 30.507 	ind 4: 0.860 	ind 5: 0.848 	ind 6: 0.845 	ind 7: 0.863 	ind 8: 0.863 	ind 9: 0.980 	ind 10: 0.916 	ind 11: 0.919 	ind 12: 1.006 	ind 13: 0.867 	ind 14: 0.900 	ind 15: 0.869 	ind 16: 0.878 	ind 17: 0.898 	ind 18: 0.893 	ind 19: 0.912 	


2021-06-15 02:30:29,853 - modnet - INFO - Generation number 7
2021-06-15 02:30:33,581 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:12<00:00,  3.13s/it]

2021-06-15 02:35:46,145 - modnet - INFO - Loss per individual: ind 0: 0.884 	ind 1: 0.917 	ind 2: 1.071 	ind 3: 0.888 	ind 4: 0.861 	ind 5: 0.875 	ind 6: 0.992 	ind 7: 0.890 	ind 8: 1.029 	ind 9: 0.884 	ind 10: 0.938 	ind 11: 0.915 	ind 12: 0.973 	ind 13: 0.851 	ind 14: 0.889 	ind 15: 0.927 	ind 16: 0.925 	ind 17: 0.932 	ind 18: 0.877 	ind 19: 0.904 	


2021-06-15 02:35:46,966 - modnet - INFO - Generation number 8
2021-06-15 02:35:50,954 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:02<00:00,  3.03s/it]

2021-06-15 02:40:53,601 - modnet - INFO - Loss per individual: ind 0: 0.902 	ind 1: 0.899 	ind 2: 0.864 	ind 3: 0.963 	ind 4: 0.900 	ind 5: 0.930 	ind 6: 0.862 	ind 7: 0.936 	ind 8: 0.870 	ind 9: 0.942 	ind 10: 0.916 	ind 11: 0.911 	ind 12: 0.920 	ind 13: 0.986 	ind 14: 0.880 	ind 15: 0.924 	ind 16: 0.915 	ind 17: 0.921 	ind 18: 0.879 	ind 19: 0.874 	


2021-06-15 02:40:54,412 - modnet - INFO - Generation number 9
2021-06-15 02:40:58,645 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:23<00:00,  3.23s/it]

2021-06-15 02:46:21,997 - modnet - INFO - Loss per individual: ind 0: 0.876 	ind 1: 0.897 	ind 2: 0.899 	ind 3: 0.857 	ind 4: 0.873 	ind 5: 0.949 	ind 6: 0.878 	ind 7: 1.000 	ind 8: 0.866 	ind 9: 0.910 	ind 10: 0.884 	ind 11: 0.896 	ind 12: 0.936 	ind 13: 0.881 	ind 14: 0.846 	ind 15: 0.900 	ind 16: 0.903 	ind 17: 0.947 	ind 18: 0.934 	ind 19: 0.900 	


2021-06-15 02:46:22,787 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


 20%|██        | 2/10 [00:00<00:00, 11.33it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


 20%|██        | 2/10 [00:00<00:00, 11.34it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:00<00:00, 10.86it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.326
  RMSE          0.693
  MDAE          0.048
  MARPD         148.366
  R2            0.457
  Correlation   0.681
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.186
  Mean-absolute Calibration Error       0.174
  Miscalibration Area                   0.176
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.237
     Group Size: 0.56 -- Calibration Error: 0.193
     Group Size: 1.00 -- Calibration Error: 0.174
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.259
     Group Size: 0.56 -- Calibration Error: 0.213
     Group Size: 1.00 -- Calibration Error: 0.186
==============

100%|██████████| 100/100 [06:45<00:00,  4.06s/it]

2021-06-15 02:53:21,192 - modnet - INFO - Loss per individual: ind 0: 1.040 	ind 1: 0.965 	ind 2: 0.957 	ind 3: 1.007 	ind 4: 0.957 	ind 5: 1.071 	ind 6: 0.903 	ind 7: 1.030 	ind 8: 0.939 	ind 9: 1.039 	ind 10: 0.962 	ind 11: 1.010 	ind 12: 1.065 	ind 13: 1.130 	ind 14: 1.171 	ind 15: 1.151 	ind 16: 1.024 	ind 17: 0.940 	ind 18: 1.083 	ind 19: 1.100 	


2021-06-15 02:53:21,987 - modnet - INFO - Generation number 1
2021-06-15 02:53:25,132 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:51<00:00,  3.51s/it]

2021-06-15 02:59:16,473 - modnet - INFO - Loss per individual: ind 0: 1.029 	ind 1: 0.979 	ind 2: 0.942 	ind 3: 0.906 	ind 4: 0.964 	ind 5: 0.973 	ind 6: 0.977 	ind 7: 0.959 	ind 8: 1.093 	ind 9: 0.991 	ind 10: 0.952 	ind 11: 1.178 	ind 12: 1.051 	ind 13: 1.009 	ind 14: 0.976 	ind 15: 1.044 	ind 16: 1.148 	ind 17: 0.990 	ind 18: 0.947 	ind 19: 1.023 	


2021-06-15 02:59:17,272 - modnet - INFO - Generation number 2
2021-06-15 02:59:20,629 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:48<00:00,  4.09s/it]

2021-06-15 03:06:09,196 - modnet - INFO - Loss per individual: ind 0: 0.964 	ind 1: 0.947 	ind 2: 1.010 	ind 3: 0.967 	ind 4: 1.152 	ind 5: 0.911 	ind 6: 0.968 	ind 7: 1.137 	ind 8: 0.927 	ind 9: 0.971 	ind 10: 1.000 	ind 11: 0.942 	ind 12: 0.920 	ind 13: 0.982 	ind 14: 0.974 	ind 15: 0.979 	ind 16: 1.047 	ind 17: 1.010 	ind 18: 0.998 	ind 19: 0.897 	


2021-06-15 03:06:10,018 - modnet - INFO - Generation number 3
2021-06-15 03:06:13,557 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:23<00:00,  3.83s/it]

2021-06-15 03:12:36,968 - modnet - INFO - Loss per individual: ind 0: 1.087 	ind 1: 0.967 	ind 2: 1.040 	ind 3: 1.043 	ind 4: 1.020 	ind 5: 1.039 	ind 6: 0.967 	ind 7: 0.963 	ind 8: 1.018 	ind 9: 1.091 	ind 10: 0.925 	ind 11: 0.972 	ind 12: 0.947 	ind 13: 0.968 	ind 14: 1.025 	ind 15: 0.991 	ind 16: 0.955 	ind 17: 0.978 	ind 18: 1.099 	ind 19: 0.946 	


2021-06-15 03:12:37,826 - modnet - INFO - Generation number 4
2021-06-15 03:12:41,424 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:44<00:00,  4.04s/it]

2021-06-15 03:19:25,556 - modnet - INFO - Loss per individual: ind 0: 0.952 	ind 1: 1.071 	ind 2: 0.996 	ind 3: 0.884 	ind 4: 1.040 	ind 5: 0.942 	ind 6: 0.965 	ind 7: 0.983 	ind 8: 0.946 	ind 9: 1.059 	ind 10: 0.967 	ind 11: 1.067 	ind 12: 0.957 	ind 13: 1.010 	ind 14: 0.949 	ind 15: 0.961 	ind 16: 1.000 	ind 17: 0.970 	ind 18: 0.924 	ind 19: 0.905 	


2021-06-15 03:19:26,390 - modnet - INFO - Generation number 5
2021-06-15 03:19:30,219 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:19<00:00,  3.80s/it]

2021-06-15 03:25:50,164 - modnet - INFO - Loss per individual: ind 0: 0.921 	ind 1: 1.130 	ind 2: 1.004 	ind 3: 1.027 	ind 4: 1.050 	ind 5: 0.990 	ind 6: 0.973 	ind 7: 0.997 	ind 8: 0.968 	ind 9: 1.004 	ind 10: 1.094 	ind 11: 1.071 	ind 12: 1.024 	ind 13: 1.061 	ind 14: 1.047 	ind 15: 0.956 	ind 16: 0.997 	ind 17: 0.971 	ind 18: 0.915 	ind 19: 0.992 	


2021-06-15 03:25:50,999 - modnet - INFO - Generation number 6
2021-06-15 03:25:55,010 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:58<00:00,  4.19s/it]

2021-06-15 03:32:53,884 - modnet - INFO - Loss per individual: ind 0: 0.908 	ind 1: 0.975 	ind 2: 1.015 	ind 3: 1.049 	ind 4: 1.071 	ind 5: 0.986 	ind 6: 1.111 	ind 7: 0.928 	ind 8: 0.961 	ind 9: 0.892 	ind 10: 0.992 	ind 11: 0.950 	ind 12: 0.965 	ind 13: 1.050 	ind 14: 0.950 	ind 15: 0.905 	ind 16: 0.983 	ind 17: 1.006 	ind 18: 0.951 	ind 19: 0.897 	


2021-06-15 03:32:54,679 - modnet - INFO - Generation number 7
2021-06-15 03:32:58,856 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [06:43<00:00,  4.04s/it]

2021-06-15 03:39:42,632 - modnet - INFO - Loss per individual: ind 0: 1.109 	ind 1: 0.899 	ind 2: 0.962 	ind 3: 1.003 	ind 4: 1.015 	ind 5: 1.001 	ind 6: 0.918 	ind 7: 0.934 	ind 8: 1.059 	ind 9: 0.951 	ind 10: 1.095 	ind 11: 1.056 	ind 12: 1.013 	ind 13: 0.940 	ind 14: 0.960 	ind 15: 0.947 	ind 16: 0.951 	ind 17: 1.010 	ind 18: 0.891 	ind 19: 0.888 	


2021-06-15 03:39:43,465 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


 20%|██        | 2/10 [00:00<00:00, 11.31it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


 20%|██        | 2/10 [00:00<00:00, 11.24it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:00<00:00, 10.77it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.274
  RMSE          0.673
  MDAE          0.027
  MARPD         151.961
  R2            0.306
  Correlation   0.586
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.118
  Mean-absolute Calibration Error       0.099
  Miscalibration Area                   0.100
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.160
     Group Size: 0.56 -- Calibration Error: 0.115
     Group Size: 1.00 -- Calibration Error: 0.099
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.179
     Group Size: 0.56 -- Calibration Error: 0.137
     Group Size: 1.00 -- Calibration Error: 0.118
==============

100%|██████████| 100/100 [04:55<00:00,  2.95s/it]

2021-06-15 03:44:50,772 - modnet - INFO - Loss per individual: ind 0: 1.278 	ind 1: 1.069 	ind 2: 1.001 	ind 3: 1.021 	ind 4: 0.976 	ind 5: 1.035 	ind 6: 1.037 	ind 7: 1.197 	ind 8: 0.974 	ind 9: 1.050 	ind 10: 1.018 	ind 11: 1.056 	ind 12: 1.049 	ind 13: 1.002 	ind 14: 1.133 	ind 15: 1.330 	ind 16: 1.218 	ind 17: 2.142 	ind 18: 0.960 	ind 19: 0.958 	


2021-06-15 03:44:51,587 - modnet - INFO - Generation number 1
2021-06-15 03:44:54,693 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:50<00:00,  2.90s/it]

2021-06-15 03:49:45,114 - modnet - INFO - Loss per individual: ind 0: 0.944 	ind 1: 0.990 	ind 2: 1.380 	ind 3: 0.988 	ind 4: 0.989 	ind 5: 1.139 	ind 6: 1.029 	ind 7: 0.989 	ind 8: 1.056 	ind 9: 1.318 	ind 10: 0.971 	ind 11: 1.051 	ind 12: 1.016 	ind 13: 1.001 	ind 14: 0.952 	ind 15: 1.013 	ind 16: 1.027 	ind 17: 0.927 	ind 18: 1.049 	ind 19: 0.998 	


2021-06-15 03:49:45,946 - modnet - INFO - Generation number 2
2021-06-15 03:49:49,129 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:34<00:00,  2.74s/it]

2021-06-15 03:54:23,288 - modnet - INFO - Loss per individual: ind 0: 1.051 	ind 1: 0.972 	ind 2: 0.947 	ind 3: 1.016 	ind 4: 1.035 	ind 5: 0.935 	ind 6: 0.987 	ind 7: 0.959 	ind 8: 0.978 	ind 9: 0.956 	ind 10: 1.088 	ind 11: 1.062 	ind 12: 1.009 	ind 13: 0.992 	ind 14: 1.023 	ind 15: 1.024 	ind 16: 1.011 	ind 17: 0.946 	ind 18: 0.949 	ind 19: 1.009 	


2021-06-15 03:54:24,125 - modnet - INFO - Generation number 3
2021-06-15 03:54:27,590 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [05:36<00:00,  3.37s/it]

2021-06-15 04:00:04,577 - modnet - INFO - Loss per individual: ind 0: 0.984 	ind 1: 1.077 	ind 2: 0.957 	ind 3: 1.025 	ind 4: 0.999 	ind 5: 0.987 	ind 6: 0.971 	ind 7: 1.001 	ind 8: 1.030 	ind 9: 1.004 	ind 10: 1.002 	ind 11: 0.973 	ind 12: 0.984 	ind 13: 0.996 	ind 14: 1.029 	ind 15: 0.980 	ind 16: 0.987 	ind 17: 1.057 	ind 18: 0.986 	ind 19: 1.055 	


2021-06-15 04:00:05,420 - modnet - INFO - Generation number 4
2021-06-15 04:00:09,018 - modnet - INFO - Multiprocessing on 10 cores. Total of 128 cores available.


100%|██████████| 100/100 [04:45<00:00,  2.85s/it]

2021-06-15 04:04:54,219 - modnet - INFO - Loss per individual: ind 0: 0.977 	ind 1: 1.004 	ind 2: 1.000 	ind 3: 0.960 	ind 4: 0.965 	ind 5: 0.955 	ind 6: 1.038 	ind 7: 0.988 	ind 8: 1.104 	ind 9: 1.008 	ind 10: 0.965 	ind 11: 0.949 	ind 12: 0.930 	ind 13: 1.031 	ind 14: 0.947 	ind 15: 1.002 	ind 16: 1.017 	ind 17: 0.975 	ind 18: 1.129 	ind 19: 1.102 	


2021-06-15 04:04:55,033 - modnet - INFO - Early stopping: same best model for 4 consecutive generations
 (1/n) Calculating accuracy metrics
 (2/n) Calculating average calibration metrics


 20%|██        | 2/10 [00:00<00:00, 11.01it/s]

 (3/n) Calculating adversarial group calibration metrics
  [1/2] for mean absolute calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


 20%|██        | 2/10 [00:00<00:00, 11.03it/s]

  [2/2] for root mean squared calibration error
Measuring adversarial group calibration by spanning group size between 0.0 and 1.0, in 10 intervals


100%|██████████| 10/10 [00:00<00:00, 10.53it/s]


 (4/n) Calculating sharpness metrics
 (n/n) Calculating proper scoring rule metrics
**Finished Calculating All Metrics**


===================== Accuracy Metrics =====================
  MAE           0.198
  RMSE          0.421
  MDAE          0.008
  MARPD         156.454
  R2            0.471
  Correlation   0.724
=============== Average Calibration Metrics ================
  Root-mean-squared Calibration Error   0.113
  Mean-absolute Calibration Error       0.098
  Miscalibration Area                   0.099
========== Adversarial Group Calibration Metrics ===========
  Mean-absolute Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.170
     Group Size: 0.56 -- Calibration Error: 0.116
     Group Size: 1.00 -- Calibration Error: 0.098
  Root-mean-squared Adversarial Group Calibration Error
     Group Size: 0.11 -- Calibration Error: 0.182
     Group Size: 0.56 -- Calibration Error: 0.134
     Group Size: 1.00 -- Calibration Error: 0.113
==============

In [15]:
maes

array([0.33880384, 0.2919107 , 0.32583968, 0.27448574, 0.19796235])

In [16]:
maes.mean()

0.2858004611596556

In [17]:
uncertainties

array([0.13591428, 0.17899893, 0.14358331, 0.13451155, 0.13358939])

In [18]:
uncertainties.mean()

0.14531949162483215

In [19]:
metricss

[{'accuracy': {'mae': 0.33880383734038305,
   'rmse': 0.6863446648592074,
   'mdae': 0.06272993981838226,
   'marpd': 148.13482658132244,
   'r2': 0.38366277611131727,
   'corr': 0.6381710485850238},
  'avg_calibration': {'rms_cal': 0.10135241924598293,
   'ma_cal': 0.08396220267188012,
   'miscal_area': 0.08480017864182088},
  'adv_group_calibration': {'ma_adv_group_cal': {'group_sizes': array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
           0.55555556, 0.66666667, 0.77777778, 0.88888889, 1.        ]),
    'adv_group_cali_mean': array([0.43801515, 0.13527645, 0.11223977, 0.10264779, 0.10101123,
           0.09757431, 0.0934084 , 0.0892574 , 0.08845365, 0.0839622 ]),
    'adv_group_cali_stderr': array([0.04633965, 0.02141775, 0.01206075, 0.0045981 , 0.00586519,
           0.00445835, 0.0036416 , 0.00254372, 0.00143897, 0.        ])},
   'rms_adv_group_cal': {'group_sizes': array([0.        , 0.11111111, 0.22222222, 0.33333333, 0.44444444,
           0.55555556, 0

#### Conclusion

Relative percentage change of the error compared to MODNet_exp_GA reference: 18%